src/result_processor.c

```diff
@@ -845,3 +846,317 @@ ResultProcessor *RPCounter_New() {
+      IsBufferEmpty(rpPufferAndLocker)) {
+    return result_status;
+  }
+
+  // Now we have the data of all documents that pass the query filters,
+  // let's lock Redis to provide safe access to Redis keyspace 
+  
+  // initialize next function to the defualt value
+  rp->Next = rpbufferNext_Yield;
+
+  // unlockspec to avoid deadlocks
+  RedisSearchCtx_UnlockSpec(RSctx);
+
+  // Lock Redis to gurentee safe access to Redis keyspace
+  LockRedis(rpPufferAndLocker, RSctx->redisCtx);
+
+  // If the spec has been changed since we released the spec lock,
+  // we need to validate every buffered result
+  if (currentIndexVersion != IndexSpec_GetVersion(RSctx->spec)) {
+    rp->Next = rpbufferNext_ValidateAndYield;
+  }
+
+  // We don't lock the index because we assume that there 
+  // are no more access to the index down the pipeline and the data 
+  // we buffered remains valid.
+  return rp->Next(rp, res);
+}
+
+bool isResultValid(const SearchResult *res) {
+  // check if the doc is not marked deleted in the spec
+  return !(res->dmd->flags & Document_Deleted);
+}
+static void InvalidateBufferedResult(SearchResult *buffered_result) {
+  buffered_result->indexResult = NULL;
+  buffered_result->dmd = NULL;
+  buffered_result->scoreExplain = NULL;
+  memset(&buffered_result->rowdata, 0, sizeof(RLookupRow));
+}
+static int ReturnResult(SearchResult *buffered_result,  SearchResult *result_output) {
+  // Free the RLookup row before overriding it.
+  RLookupRow_Cleanup(&result_output->rowdata);
+  *result_output = *buffered_result;
+
+  InvalidateBufferedResult(buffered_result);
+  return RS_RESULT_OK;
+}
+/*********** Redis lock management ***********/ 
+bool isRedisLocked(RPBufferAndLocker *bufferAndLocker) {
+  return bufferAndLocker->isRedisLocked;
+}
+
+void LockRedis(RPBufferAndLocker *rpBufferAndLocker, RedisModuleCtx* redisCtx) {
+  RedisModule_ThreadSafeContextLock(redisCtx);
+
+  rpBufferAndLocker->isRedisLocked = true;
+}
+
+void UnLockRedis(RPBufferAndLocker *rpBufferAndLocker, RedisModuleCtx* redisCtx) {
+  RedisModule_ThreadSafeContextUnlock(redisCtx);
+
+  rpBufferAndLocker->isRedisLocked = false;
+}
+/*********** Yeild results phase functions ***********/ 
+
+int rpbufferNext_Yield(ResultProcessor *rp, SearchResult *result_output) {
+  RPBufferAndLocker *RPBuffer = (RPBufferAndLocker *)rp;
+  SearchResult *curr_res = GetNextResult(RPBuffer);
+  
+  if(!curr_res) {
+    return RS_RESULT_EOF;
+  }
+ return ReturnResult(curr_res, result_output);
+  
+}
+
+int rpbufferNext_ValidateAndYield(ResultProcessor *rp, SearchResult *result_output) {
+  RPBufferAndLocker *RPBuffer = (RPBufferAndLocker *)rp;
+  SearchResult *curr_res = NULL;
+  
+  // iterate the buffer.
+  while(curr_res = GetNextResult(RPBuffer)) {
+    // Skip invalid results
+    if (isResultValid(curr_res)) {
+      return ReturnResult(curr_res, result_output);
+    }
+
+    // If the result is invalid discard it.
+    SearchResult_Clear(curr_res);
```

Comment:
The only difference is that `Destroy` also frees the `dyn` array which Im not sure that there is a scenario where it exists at this point because it is created when we load the documents

Reply:
It can be created in the metric loader